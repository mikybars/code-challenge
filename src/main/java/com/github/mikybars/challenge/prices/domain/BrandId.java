package com.github.mikybars.challenge.prices.domain;

public record BrandId(String id) {

  public static BrandId zara() {
    return new BrandId("1");
  }
}
