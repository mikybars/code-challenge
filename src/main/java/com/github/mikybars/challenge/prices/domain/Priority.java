package com.github.mikybars.challenge.prices.domain;

public record Priority(int priority) implements Comparable<Priority> {

  @Override
  public int compareTo(Priority other) {
    return Integer.compare(priority, other.priority);
  }
}
