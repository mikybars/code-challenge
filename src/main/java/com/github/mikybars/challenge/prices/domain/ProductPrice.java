package com.github.mikybars.challenge.prices.domain;

import java.time.LocalDateTime;

public record ProductPrice(ProductId productId, BrandId brandId, PriceListId priceListId,
                           LocalDateTime startDate, LocalDateTime endDate,
                           Priority priority, Money price) {

}
