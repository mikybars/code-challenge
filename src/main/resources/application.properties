spring.application.name=code-challenge
spring.mvc.problemdetails.enabled=true
