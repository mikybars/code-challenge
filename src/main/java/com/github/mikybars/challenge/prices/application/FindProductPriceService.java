package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.prices.application.ports.in.FindProductPriceUseCase;
import com.github.mikybars.challenge.prices.domain.BrandId;
import com.github.mikybars.challenge.prices.domain.ProductId;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.time.LocalDateTime;
import org.springframework.stereotype.Service;

@Service
class FindProductPriceService implements FindProductPriceUseCase {

  @Override
  public ProductPrice execute(LocalDateTime applicationDate, ProductId productId, BrandId brandId) {
    return null;
  }
}
