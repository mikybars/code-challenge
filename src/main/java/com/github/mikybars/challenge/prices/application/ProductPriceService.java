package com.github.mikybars.challenge.prices.application;

import com.github.mikybars.challenge.common.NotFoundException;
import com.github.mikybars.challenge.prices.application.ports.in.GetProductPriceUseCase;
import com.github.mikybars.challenge.prices.application.ports.out.ProductPriceRepository;
import com.github.mikybars.challenge.prices.domain.BrandId;
import com.github.mikybars.challenge.prices.domain.ProductId;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.time.LocalDateTime;
import java.util.function.Supplier;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
class ProductPriceService implements GetProductPriceUseCase {

  private final ProductPriceRepository productPriceRepository;

  @Override
  public ProductPrice execute(LocalDateTime applicationDate, ProductId productId, BrandId brandId) {
    return productPriceRepository
        .findTheHighestRankedBy(applicationDate, productId, brandId)
        .orElseThrow(notFoundException(applicationDate, productId, brandId));
  }

  private static Supplier<NotFoundException> notFoundException(
      LocalDateTime applicationDate, ProductId productId, BrandId brandId) {
    return () -> new NotFoundException(
        "no product price found for [applicationDate=%s, productId=%s, brandId=%s]".formatted(
            applicationDate, productId.id(), brandId.id()));
  }
}
