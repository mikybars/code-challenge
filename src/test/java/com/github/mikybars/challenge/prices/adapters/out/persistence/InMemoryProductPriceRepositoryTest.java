package com.github.mikybars.challenge.prices.adapters.out.persistence;

import static org.assertj.core.api.Assertions.assertThat;

import com.github.mikybars.challenge.JpaConfiguration;
import com.github.mikybars.challenge.prices.application.ProductPriceSearchCriteria;
import com.github.mikybars.challenge.prices.domain.BrandId;
import com.github.mikybars.challenge.prices.domain.ProductId;
import com.github.mikybars.challenge.prices.domain.ProductPrice;
import java.time.LocalDateTime;
import java.util.Optional;
import org.approvaltests.JsonApprovals;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.context.annotation.Import;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;

@DataJpaTest
@ActiveProfiles("standalone")
@Import({
    JpaConfiguration.class,
    InMemoryProductPriceRepository.class,
    ProductPricePersistenceMapperImpl.class
})
@Sql("classpath:prices.sql")
class InMemoryProductPriceRepositoryTest {

  static final LocalDateTime onlyPriceStartDate = LocalDateTime.parse("2023-01-01T00:00:00");
  static final LocalDateTime onlyPriceEndDate = LocalDateTime.parse("2023-12-31T23:59:59");
  static final LocalDateTime highestRankedStartDate = LocalDateTime.parse("2023-11-20T00:00:00");
  static final ProductId productWithOnePrice = new ProductId("p1");
  static final BrandId brandWithOnePrice = new BrandId("b1");

  @Autowired
  InMemoryProductPriceRepository productPriceRepository;

  @Test
  void returnEmptyWhenDateIsBefore() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate.minusSeconds(1), productWithOnePrice,
            brandWithOnePrice));

    assertThat(productPrice).isEmpty();
  }

  @Test
  void returnOneWhenDateEqualsStart() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate, productWithOnePrice, brandWithOnePrice));

    assertThat(productPrice).get()
        .satisfies(JsonApprovals::verifyAsJson);
  }

  @Test
  void returnOneWhenDateIsInBetween() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceStartDate.plusSeconds(1), productWithOnePrice,
            brandWithOnePrice));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId)
        .containsExactly(productWithOnePrice, brandWithOnePrice);
  }

  @Test
  void returnOneWhenDateEqualsEnd() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceEndDate, productWithOnePrice, brandWithOnePrice));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId)
        .containsExactly(productWithOnePrice, brandWithOnePrice);
  }

  @Test
  void returnEmptyWhenDateIsAfter() {
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(onlyPriceEndDate.plusSeconds(1), productWithOnePrice,
            brandWithOnePrice));

    assertThat(productPrice).isEmpty();
  }

  @Test
  void returnHighestRanked() {
    ProductId productWithMultiplePrices = new ProductId("p2");
    BrandId brandWithMultiplePrices = new BrandId("b1");
    Optional<ProductPrice> productPrice = productPriceRepository.findTheHighestRankedBy(
        new ProductPriceSearchCriteria(highestRankedStartDate, productWithMultiplePrices,
            brandWithMultiplePrices));

    assertThat(productPrice).get()
        .extracting(ProductPrice::productId, ProductPrice::brandId, ProductPrice::startDate)
        .containsExactly(productWithMultiplePrices, brandWithMultiplePrices, highestRankedStartDate);
  }
}