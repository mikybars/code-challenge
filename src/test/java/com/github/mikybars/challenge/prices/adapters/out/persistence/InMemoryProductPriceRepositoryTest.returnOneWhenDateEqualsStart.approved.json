{
  "productId": {
    "id": "p1"
  },
  "brandId": {
    "id": "b1"
  },
  "priceListId": {
    "id": "pl1"
  },
  "startDate": "2023-01-01T00:00",
  "endDate": "2023-12-31T23:59:59",
  "price": {
    "amount": 100.00,
    "currency": "USD"
  }
}